vms/platformvm/txs/fee/calculator.go

```diff
@@ -4,85 +4,172 @@
-		upgradeTimes: upgradeTimes,
-		time:         chainTime,
+		c: &calculator{
+			upgrades:  upgradeTimes,
+			staticCfg: config,
+			time:      chainTime,
+		},
+	}
+}
+
+// NewDynamicCalculator must be used post E upgrade activation
+func NewDynamicCalculator(
+	cfg StaticConfig,
+	feeManager *fees.Manager,
+	blockMaxComplexity fees.Dimensions,
+) *Calculator {
+	return &Calculator{
+		c: &calculator{
+			isEActive:          true,
+			staticCfg:          cfg,
+			feeManager:         feeManager,
+			blockMaxComplexity: blockMaxComplexity,
+			// credentials are set when CalculateFee is called
+		},
 	}
 }
 
 type Calculator struct {
-	config       StaticConfig
-	upgradeTimes upgrade.Config
-	time         time.Time
+	c *calculator
 }
 
-// [CalculateFee] returns the minimal fee needed to accept [tx], at chain time [time]
-func (c *Calculator) CalculateFee(tx txs.UnsignedTx) uint64 {
-	tmp := &calculator{
-		upgrades:  c.upgradeTimes,
-		staticCfg: c.config,
-		time:      c.time,
-	}
+func (c *Calculator) GetFee() uint64 {
+	return c.c.fee
+}
+
+func (c *Calculator) ResetFee(newFee uint64) {
+	c.c.fee = newFee
+}
+
+func (c *Calculator) ComputeFee(tx txs.UnsignedTx, creds []verify.Verifiable) (uint64, error) {
+	c.c.credentials = creds
+	c.c.fee = 0 // zero fee among different ComputeFee invocations (unlike Complexity which gets cumulated)
+	err := tx.Visit(c.c)
+	return c.c.fee, err
+}
+
+func (c *Calculator) AddFeesFor(complexity fees.Dimensions) (uint64, error) {
+	return c.c.addFeesFor(complexity)
+}
+
+func (c *Calculator) RemoveFeesFor(unitsToRm fees.Dimensions) (uint64, error) {
+	return c.c.removeFeesFor(unitsToRm)
+}
 
-	// this is guaranteed to never return an error
-	_ = tx.Visit(tmp)
-	return tmp.fee
+func (c *Calculator) GetCumulatedComplexity() fees.Dimensions {
+	if c.c.feeManager != nil {
+		return c.c.feeManager.GetCumulatedComplexity()
+	}
+	return fees.Empty
 }
 
-// calculator is intentionally unexported and used through Calculator to provide
-// a more convenient API
 type calculator struct {
-	// Pre E-fork inputs
-	upgrades  upgrade.Config
+	// setup
+	isEActive bool
 	staticCfg StaticConfig
-	time      time.Time
+
+	// Pre E-upgrade inputs
+	upgrades upgrade.Config
+	time     time.Time
+
+	// Post E-upgrade inputs
+	feeManager         *fees.Manager
+	blockMaxComplexity fees.Dimensions
+	credentials        []verify.Verifiable
 
 	// outputs of visitor execution
 	fee uint64
 }
 
 func (c *calculator) AddValidatorTx(*txs.AddValidatorTx) error {
+	// AddValidatorTx is banned following Durango activation, so we
+	// only return the pre EUpgrade fee here
 	c.fee = c.staticCfg.AddPrimaryNetworkValidatorFee
 	return nil
 }
 
-func (c *calculator) AddSubnetValidatorTx(*txs.AddSubnetValidatorTx) error {
-	c.fee = c.staticCfg.AddSubnetValidatorFee
-	return nil
+func (c *calculator) AddSubnetValidatorTx(tx *txs.AddSubnetValidatorTx) error {
+	if !c.isEActive {
+		c.fee = c.staticCfg.AddSubnetValidatorFee
+		return nil
+	}
```

Comment:
I think these `if` checks are going against the refactor we're trying to do where we're having separate code paths for static + dynamic fees. Maybe we can make a new calculator struct that has the dynamic fees that the `DynamicFeeCalculator` will wrap around, while the `StaticFeeCalculator` will continue wrapping around this old struct?

Reply:
Done